### ISSUE TITLE
Specify strict order of evaluation for a subexpression 

### ISSUE THREAD

**Issue by dnovatchev** [qt4cg/qtspecs#78]
As discussed in a related issue #71, given an XPath expression such as (1):

```
for $d in  ( 10, 2, 3 , current-date())
  return
     $d[. castable as xs:date][xs:date(.) le current-date()]
```
anyone who expects this expression to be evaluated without errors and to produce as result a sequence of one item, will be disappointed to get an error (as per BaseX 9.5.2):

**Error**:
_Stopped at C:/W3C-XPath/DupsSolutions/file, 3/39:
[XPTY0004] Cannot convert xs:integer to `xs:date`: 10_.

At present the recommended solution to the problem is to write an expression of this kind instead (2):

```
for $d in  ( 10, 2, 3, current-date() )
  return
     if($d[. castable as xs:date] eq $d)
       then $d[xs:date(.) le current-date()]
       else()
```

Evaluating this produces the expected result (a sequence of one item, which is the `current-date()` ).

There are many challenges with such a recommendation:

1. The expression above is unreadable.
2. It is very difficult and error-prone to convert manually (1) to (2)
3. It would be nearly impossible to transform a more complex expression and it would be tremendously difficult to read, understand and maintain such code.

**Proposed solution**:

 Introduce the **_strict-order evaluation_** operator **~**

Then achieving a strict-order evaluation for the subexpression (of (1) above): `[xs:date(.) le current-date()]` would be  simply (3):

```
for $d in  ( 10, 2, 3 , current-date())
  return
     $d[. castable as xs:date]~[xs:date(.) le current-date()]
```

In this particular case the XPath processor will rewrite the above expression into:

```
for $d in  ( 10, 2, 3, current-date() )
  return
     $d[. castable as xs:date] => 
                                  (function($x) {
                                                 $x[xs:date($x) le current-date()]
                                                 }
                                  ) ()
```

**Comment by AlainCouthures** [comment]
Would `$d[. castable as xs:date]!.[xs:date(.) le current-date()]` be a good equivalent?

**Comment by ChristianGruen** [comment]
I have changed the behavior of BaseX as described in https://github.com/qt4cg/qtspecs/issues/71#issuecomment-850955440; this way, it shouldn’t make a difference anymore if values that are filtered by predicates are static or dynamic.

Regarding the strict order of predicates, I’m favoring the original proposal made in #71.

**Comment by benibela** [comment]
Perhaps there should be monads for proper ordering

**Comment by dnovatchev** [comment]
@AlainCouthures :
> $d[. castable as xs:date]!.[xs:date(.) le current-date()]

Unfortunately, this raises an error:

![image](https://user-images.githubusercontent.com/10605892/120113408-36af1680-c12f-11eb-9662-983c8684115b.png)

**Comment by ChristianGruen** [comment]
Once again, this error is a result of the static evaluation of the expression. BaseX 9.5.2 has just been released (thanks for checking it out), but it might be worth checking the [even newer snapshot](https://files.basex.org/releases/latest/). It includes the revised evaluation that I described, and which should be more intuitive.

**Comment by dnovatchev** [comment]
@ChristianGruen, Re: 
> Regarding the strict order of predicates, I’m favoring the original proposal made in #71.

I initially thought that #71 was editorial and proposed to correct the example in the Spec. Now I see that it actually proposes to ban any reordering at all...

My standpoint on this is that #71 is replacing one extreme with another -- following the "All or Nothing" principle usually leads to inflexible results. If the other extreme has been accepted in the past, and is still the official rule, maybe there was quite some rationale in doing so?

This proposal gives the XPath developer full control without banning anything: leave possible reordering whenever this makes sense or we don't care, but be in control and specify strict order exactly in the cases when this matters to us.

**Comment by ChristianGruen** [comment]
I’m not sure if it was an extreme choice (Michael might be more competent in judging this). I would rather guess that most implementations evaluated predicates in the natural order anyway – such as BaseX does, too.

I would even vote for enforcing evaluation order for logical expressions, as this is the behavior that people are used to from most other programming languages.

**Comment by dnovatchev** [comment]
> I would even vote for enforcing evaluation order for logical expressions, as this is the behavior that people are used to from most other programming languages.

If by this you mean that we need to have rules for **_shortcutting_**,  such as:

`false() and $anyExpr` must produce `false()`

then yes, I am also for such shortcutting as provided in most programming languages.

This proposal goes just a little bit further, to give the programmer, besides the shortcutting rules, explicit control to specify order of evaluation in any other context.

**Comment by michaelhkay** [comment]
If the proposed "~" is an operator, it's not clear what its operands are. I think it's more of a pseudo-operator, a syntactic marker.

I think I would prefer to avoid new syntax. I think we can require a strict order-of-evaluation where multiple predicates are used (`a[p][q]`) because users who want to allow reordering can use the alternative `a[p and q]`. Implementations that currently do reordering of predicates could provide a configuration option to permit such behaviour, but reordering would be disallowed by default.

For contexts other than predicates, I'd like to see strictly-shortcutting versions of `and` and `or`, for example `and-also` and `or-else`, where `a and-also b` equates to `if (a) then b else false()`, e.g `if ($d castable as xs:date and-also year-from-date(xs:date($d)) = 2010)`.

Alternative names for the shortcutting operators can be found at https://en.wikipedia.org/wiki/Short-circuit_evaluation - for example "and if", "and then", "and also" are all in use, with various punctuation.

**Comment by ChristianGruen** [comment]
@michaelhkay Why do you think that the evaluation order of the existing logical expressions cannot be made strict, too?

**Comment by michaelhkay** [comment]
Because some XQuery vendors translate operators such as and/or to other languages such as SQL, and reordering of terms in a WHERE clause to take maximum advantage of indexes is a key feature of SQL optimisers.

**Comment by dnovatchev** [comment]
@michaelhkay:
> If the proposed "~" is an operator, it's not clear what its operands are.

The operands are:

- LHS (1st argument): any subexpression produced by the first non-terminal symbol in the RHS of the grammar rule used to parse the expression. 
- RHS (2nd argument): any subexpression produced by a following (not the 1st) non-terminal symbol in the RHS of the grammar rule used to parse the expression.

I think that there are existing tools, that given a context-free grammar can find the sets of all such **_leadingNT_** s and **_followingNT_** s (we actually need just the set of the _followingNT_ s)

When these are identified, the grammar can be updated to insert `~?`  before any of the **_followingNT_**s

**Comment by michaelhkay** [comment]
"The operands are: ... LHS (1st argument): any subexpression produced by the first non-terminal symbol..."

But in your example the "operands" in this sense are predicates in square brackets, e.g. `[$x castable as xs:date]`, and predicates are not expressions. For it to be a true first-order operator, the grammatical subclauses must be expressions, and the semantics must involve evaluating those expressions to product an XDM value, then combining the XDM values in some way. There are also higher-order operators such as "!" and "x[y]" where the subclauses are expressions but the result is not obtained simply by evaluating the subexpressions and combining their results; and there are pseudo-operators such as "instance of" where the subclauses are not expressions. I think your proposed "~" is in this sense a pseudo-operator.

**Comment by dnovatchev** [comment]
> But in your example the "operands" in this sense are predicates in square brackets, e.g. `[$x castable as xs:date]`, and predicates are not expressions.

The 1st "operand" / argument is everything to the left of `~` that has been already read and parsed until encountering the `~` operator. This has to be an expression. The second operand is (whatever is produced by parsing) the concatenation of all non-terminal and terminal symbols following `~` in this rule. The precedence of `~` must be the lowest in the precedence table.

> I think your proposed "~" is in this sense a pseudo-operator.

No arguing how we can best classify this, what is important here is its power and usefulness. We could call it a "marker" or even a "hint" to the compiler/interpreter.

**Comment by rhdunn** [comment]
This is something that could be handled by a pragma [1], e.g. `$x[ (# strict-order #) { . castable as xs:date } ][ xs:date(.) le current-date() ]`. That has the advantages that:
1. it is an expression;
2. parsing pragmas is well-defined;
3. the prama syntax is already supported in XQuery (and XPath with the Full Text extension).

The XQuery/XPath 4.0 specifications could define a `strict-order` pragma (or other name) that indicates that the evaluation of the expression should not be re-ordered. If it is used in a predicate as the entirety of that predicate, then the no re-ordering logic applies to the predicate as a whole. If the predicate occurs in a logic expression, the expression must be evaluated before the expressions that follow.

[1] https://www.w3.org/TR/xquery-31/#id-extension-expressions

**Comment by michaelhkay** [comment]
We have resolved this: see section §2.4.5 of the XPath spec on guarded expressions.
